### PR TITLE
Updates to allow override templates to all types of form controls

### PR DESCRIPTION
**Pull request by amrutha-rajiv**
## **Description**
- ControlTemplates house templates for all controls
- templates can be overriden via templates added to form
- custom controls can be added using ng-templates as well
#### **Verify that...**

- [x] Any related demos where added and `npm start` still works
- [x] New demos work in `Safari`, `Chrome` and `Firefox`
- [x] `npm run lint` passes
- [x] `npm test` passes and code coverage is increased
- [x] `npm run compile` still works

##### **Screenshots**

### CODE REVIEW

File: src/platform/elements/form/DynamicForm.ts [hunk 1]
```diff
@@ -80,129 +55,139 @@ export class NovoFieldsetElement {
...
+        this.form.fieldsets = [...this.fieldsets];
+    }
 
-        // Hide required fields that have been successfully filled out
-        if (hideRequiredWithValue && !Helpers.isBlank(this.form.value[control.key]) && (!control.isEmpty || control.isEmpty && control.isEmpty(this.form.controls[control.key]))) {
-          this.form.controls[control.key].hidden = true;
+    ngAfterContentInit() {
+        if (this.customTemplates && this.customTemplates.length) {
+            this.customTemplates.forEach((template: any) => {
+                this.templates.addCustom(template.type, template.template);
```
Comment (bvkimball) [inline src/platform/elements/form/DynamicForm.ts]: Wouldn't this make defining the component on one form add it to all forms?

I would think we would want to add them independently?

Comment (amrutha-rajiv) [reply src/platform/elements/form/DynamicForm.ts]: Yea, I was thinking the same. I think you should be able to do both. App level overrides/custom templates and form level overrides. What do you think?

**Comment by jgodi** [comment]
Can you resolve the spacing issues? There is an editor-config file and there are extensions for IDEs to read that in and update he spacing settings based on that. 

Will help moving from project to project where the spacing could be different.

File: src/app/pages/elements/form/FormDemo.ts [hunk 2]
```diff
@@ -387,7 +387,7 @@ export class FormDemoComponent {
       { token: 'TOKEN', military: true },
       {
         customfield: {
-          customControl: CustomDemoComponent,
+          customControl: 'custom-demo-component',
```
Comment (jgodi) [inline src/app/pages/elements/form/FormDemo.ts]: Can this be renamed to `template` just to align with tables? They pass `template` to the column config to specify a custom cell.

File: src/app/pages/elements/form/templates/VerticalDynamicForm.html [hunk 3]
```diff
@@ -1,6 +1,10 @@
 <button theme="secondary" (click)="nomyform.showAllFields()">Show All Fields</button>
 <button theme="secondary" (click)="nomyform.showOnlyRequired()">Show Required Fields</button>
-<novo-dynamic-form layout="vertical" class="dynamic" [controls]="dynamicVertical" [(form)]="dynamicVerticalForm" #nomyform></novo-dynamic-form>
+<novo-dynamic-form layout="vertical" class="dynamic" [controls]="dynamicVertical" [(form)]="dynamicVerticalForm" #nomyform>
+  <ng-template novoTemplate="custom-demo-component" type="custom-demo-component" let-control let-form="form">
```
Comment (jgodi) [inline src/app/pages/elements/form/templates/VerticalDynamicForm.html]: What does the `type` do? That is already passed in the `novoTemplate` right? The value is the type.

File: src/platform/elements/form/Control.ts [hunk 4]
```diff
@@ -234,372 +136,396 @@ export class NovoCustomControlContainerElement {
...
-        this.fieldInteractionApi.form = this.form;
-        this.fieldInteractionApi.currentKey = this.control.key;
-        try {
-          interaction.script(this.fieldInteractionApi, this.control.key);
-        } catch (err) {
-          console.info('Field Interaction Error!', this.control.key); // tslint:disable-line
-          console.error(err); // tslint:disable-line
+        this.templateContext.$implicit.tooltipPosition = this.tooltipPosition;
+        this.templateContext.$implicit.tooltip = this.tooltip;
+        this.templateContext.$implicit.startupFocus = this.control.startupFocus;//what does this do????
```
Comment (jgodi) [inline src/platform/elements/form/Control.ts]: This should focus the first field when the form is loaded, should see it on the form pages and should be autofocused

File: src/platform/elements/form/Control.ts [hunk 5]
```diff
@@ -234,372 +136,396 @@ export class NovoCustomControlContainerElement {
...
+        this.templateContext.$implicit.fileBrowserImageUploadUrl = this.control.fileBrowserImageUploadUrl;//what does this do????
```
Comment (jgodi) [inline src/platform/elements/form/Control.ts]: Added for CKEditor, so images can be uploaded

File: src/platform/elements/form/Control.ts [hunk 6]
```diff
@@ -234,372 +136,396 @@ export class NovoCustomControlContainerElement {
...
+        this.templateContext.$implicit.minimal = this.control.minimal;//what does this do????
```
Comment (jgodi) [inline src/platform/elements/form/Control.ts]: Used in the grouped forms for minimal styling (padding/label/etc)

File: src/platform/elements/form/Control.ts [hunk 7]
```diff
@@ -234,372 +136,396 @@ export class NovoCustomControlContainerElement {
...
+        this.templateContext.$implicit.config = this.control.config;// IS THIS NEEDED???
```
Comment (jgodi) [inline src/platform/elements/form/Control.ts]: Not sure if it is

File: src/app/pages/elements/form/FormDemo.ts (hunk 2 above)
Comment (amrutha-rajiv) [reply src/app/pages/elements/form/FormDemo.ts]: sure

Comment (amrutha-rajiv) [reply src/app/pages/elements/form/FormDemo.ts]: done

File: src/app/pages/elements/form/templates/VerticalDynamicForm.html (hunk 3 above)
Comment (amrutha-rajiv) [reply src/app/pages/elements/form/templates/VerticalDynamicForm.html]: type has been removed now

**Pull request merged**